src/qibo/backends/numpy.py

```diff
@@ -193,9 +193,26 @@ def transpose_state(self, pieces, state, nqubits, order): # pragma: no cover
     def random_uniform(self, shape, dtype='DTYPE'):
         return self.backend.random.random(shape).astype(self.dtypes(dtype))
 
-    def sample_measurements(self, probs, nshots):
+    def sample_shots(self, probs, nshots):
         return self.np.random.choice(range(len(probs)), size=nshots, p=probs)
 
+    def update_frequencies(self, probs, nsamples, frequencies):
+        samples = self.sample_shots(probs, nsamples)
+        res, counts = self.unique(samples, return_counts=True)
+        frequencies[res] += counts
+        return frequencies
+
+    def sample_frequencies(self, probs, nshots):
+        from qibo.config import SHOT_BATCH_SIZE
+        n = int(probs.shape[0])
+        frequencies = self.zeros(n, dtype=self.dtypes('DTYPEINT'))
+        for _ in range(nshots // SHOT_BATCH_SIZE):
```

Comment:
Are you sure this for loop is needed? Looks like you are running this nrep times and the calling the categorical nrep times. This is the big difference with your pure tf code, where there is just a single for loop.

Reply:
Being a little be more specific, the `sample_frequencies` function is called several times in your circuit example, this looks like the main difference when comparison to your pure tf code.

Reply:
I think the loop is needed otherwise the frequencies will not correspond to `nshots` measurements. Particularly the sum of all frequencies should be `nshots` which is the case now but would not be if this loop is removed. Also I think `tf.random.categorical` is called once for every loop repetition because `self.sample_shots` is called with `nshots=SHOTS_BATCH_SIZE` so the inner loop corresponds to just `range(1)` (so there is no really nested loop happening).

On the other hand I agree that there is a discrepancy somewhere in the code that I am missing right now because according to the tf-only example the GPU should be faster than CPU.

Reply:
Indeed, other observations, the `self.unique` seems to take 20% of the overall time, but the remaining comes from the `sample_shots`. 

I think your pure tf code does not forces tensorflow to copy data back to CPU. Please try to append the results to a `sample` list and then print e.g. `sample[0]` before the timing computation (remember this behaviour when we have been playing with tf.function).

We have to find a mechanism to reduce data transfers.

Reply:
(maybe making the transfer async)

Reply:
I modified `sample_frequencies` to not use `sample_shots` so that we don't have any issue with the nested for loop. From a tf perspective now the code is equivalent to:

```Python
import time
import tensorflow as tf

def sample_frequencies(probs, nshots, batch=1000000):
    """Samples measurement frequencies given a probability distribution."""
    logits = tf.math.log(probs)[tf.newaxis]
    def update_frequencies(logits, nsamples, frequencies):
        samples = tf.random.categorical(logits, nsamples, dtype=tf.int64)[0]
        res, idx, counts = tf.unique_with_counts(samples, out_idx=tf.int64)
        return tf.tensor_scatter_nd_add(frequencies, res[:, tf.newaxis], counts)

    frequencies = tf.zeros(int(probs.shape[0]), dtype=tf.int64)
    start_time = time.time()
    for _ in range(nshots // batch):
        frequencies = update_frequencies(batch, frequencies)
    frequencies = update_frequencies(nshots % batch, frequencies)
    print(time.time() - start_time)
    return frequencies
```

@scarrazza do you have any example that this Qibo-free code runs faster on GPU? For me for probs of size (1024,) (10 qubits) and nshots between 1e8-5e8 it is faster in my notebook CPU than the DGX. However I suspect that the reason for this is that the DGX is full of processes because the times I get there show large variations.

Regarding data transfer, do you think there is any transfer happening in this? I thought that since the whole function uses tf-primitives everything should run on GPU without any transfer.

Reply:
@stavros11 thanks. All examples, involving the tf.random.categorical are faster on CPU, on both setups Qibo and pure tf, including your latest commit.

I the bottleneck could be at the level of communication, given that for a single batch evaluation the performance on CPU and GPU are similar.

Reply:
> I the bottleneck could be at the level of communication, given that for a single batch evaluation the performance on CPU and GPU are similar.

Ideally for the frequency calculation there should be no communication in the sense that we don't want to transfer the big tensor of samples on the CPU. When the next batch is calculated, the previous batch should be discarded completely (that's why the same variable `samples` is used). I am not sure if Tensorflow does exactly that though.

> @stavros11 thanks. All examples, involving the tf.random.categorical are faster on CPU, on both setups Qibo and pure tf, including your latest commit.

An alternative to categorical would be to follow the method Carlos proposed. A way to do this with tf primitives is the following:
```Python
def update_frequencies(probs, nsamples, frequencies):
    cumprobs = tf.math.cumsum(probs)
    cumprobs = tf.repeat(cumprobs[tf.newaxis], nsamples, axis=0)
    samples = tf.random.uniform((nsamples,), dtype=tf.float64)
    samples = tf.math.greater(samples[:, tf.newaxis], cumprobs)
    samples = tf.reduce_sum(tf.cast(samples, dtype=tf.int64), axis=1)
    res, idx, counts = tf.unique_with_counts(samples, out_idx=tf.int64)
    return tf.tensor_scatter_nd_add(frequencies, res[:, tf.newaxis], counts)
```
The main disadvantage is that `cumprobs` will become a tensor of shape (batch size, 2^nqubits) so this will require to reduce the batch size. This is much slower than categorical on CPU but I am not sure if it could help on GPU.

Alternatively we could do a custom operator that combines the use of sampling + unique for this example.

Reply:
I am trying the current code with your test code on GPU (12GB) with 20 qubits but I get an OOM with the default batch size. I am just wondering if the performance we see on GPU is due to a "small" circuit. If so, then we should just benchmark the alternative and see what is the most efficient.

Reply:
> I am trying the current code with your test code on GPU (12GB) with 20 qubits but I get an OOM with the default batch size. I am just wondering if the performance we see on GPU is due to a "small" circuit. 

I am not sure if the circuit size matters so much because I am only measuring sampling time (not execution) but indeed the GPU may be faster than CPU for 20 qubits. Is that the case for smaller batch size? Unfortunately I cannot test this on DGX because if there are other processes running the timings do not make much sense.
Regarding the batch size, perhaps we should find a better way to set the default value because the optimal value that does not go OOM depends both on the GPU memory and the circuit size.

> If so, then we should just benchmark the alternative and see what is the most efficient.

The alternative I wrote above is much slower and gets OOM with smaller batch sizes, at least in my notebook. The only alternative that I can think of right now and could help would be doing the following in C++:
```C++
for (int i = 0; i < nsamples; i++) {
    for (long j = 0; j < nstates; j++) {
        if (random() < cumprobs[j]) {
            frequencies[j]++;
            break;
    }
}
```
where `random()` is just uniform in [0, 1]. This does the frequency calculation directly without generating any samples and calling unique, so we also don't need to specify any batch size and won't have OOM issues.